polkadot/node/core/prospective-parachains/src/fragment_tree.rs

```diff
@@ -763,43 +763,106 @@ impl FragmentTree {
-	pub(crate) fn select_child(
+	pub(crate) fn select_children(
 		&self,
 		required_path: &[CandidateHash],
+		count: u32,
 		pred: impl Fn(&CandidateHash) -> bool,
-	) -> Option<CandidateHash> {
+	) -> Vec<CandidateHash> {
 		let base_node = {
 			// traverse the required path.
 			let mut node = NodePointer::Root;
 			for required_step in required_path {
-				node = self.node_candidate_child(node, &required_step)?;
+				if let Some(next_node) = self.node_candidate_child(node, &required_step) {
+					node = next_node;
+				} else {
+					return vec![]
+				};
 			}
 
 			node
 		};
 
-		// TODO [now]: taking the first selection might introduce bias
+		// TODO: taking the first best selection might introduce bias
 		// or become gameable.
 		//
 		// For plausibly unique parachains, this shouldn't matter much.
 		// figure out alternative selection criteria?
-		match base_node {
+		self.select_children_inner(base_node, count, count, &pred, &mut vec![])
+	}
+
+	// Try finding a candidate chain starting from `base_node` of length `expected_count`.
+	// If not possible, return the longest we could find.
+	// Does a depth-first search, since we're optimistic that there won't be more than one such
+	// chains. Assumes that the chain must be acyclic.
+	fn select_children_inner(
+		&self,
+		base_node: NodePointer,
+		expected_count: u32,
+		remaining_count: u32,
+		pred: &dyn Fn(&CandidateHash) -> bool,
+		accumulator: &mut Vec<CandidateHash>,
+	) -> Vec<CandidateHash> {
+		if remaining_count == 0 {
+			// The best option is the chain we've accumulated so far.
+			return accumulator.to_vec();
+		}
+
+		let children: Vec<_> = match base_node {
 			NodePointer::Root => self
 				.nodes
 				.iter()
-				.take_while(|n| n.parent == NodePointer::Root)
-				.filter(|n| self.scope.get_pending_availability(&n.candidate_hash).is_none())
-				.filter(|n| pred(&n.candidate_hash))
-				.map(|n| n.candidate_hash)
-				.next(),
-			NodePointer::Storage(ptr) => self.nodes[ptr]
-				.children
-				.iter()
-				.filter(|n| self.scope.get_pending_availability(&n.1).is_none())
-				.filter(|n| pred(&n.1))
-				.map(|n| n.1)
-				.next(),
+				.enumerate()
+				.take_while(|(_, n)| n.parent == NodePointer::Root)
+				.filter(|(_, n)| self.scope.get_pending_availability(&n.candidate_hash).is_none())
+				.filter(|(_, n)| pred(&n.candidate_hash))
+				.map(|(ptr, n)| (NodePointer::Storage(ptr), n.candidate_hash))
+				.collect(),
+			NodePointer::Storage(base_node_ptr) => {
+				let base_node = &self.nodes[base_node_ptr];
+
+				base_node
+					.children
+					.iter()
+					.filter(|(_, hash)| self.scope.get_pending_availability(&hash).is_none())
+					.filter(|(_, hash)| pred(&hash))
+					.map(|(ptr, hash)| (*ptr, *hash))
+					.collect()
+			},
+		};
+
+		let mut best_result = accumulator.clone();
+		for (child_ptr, child_hash) in children {
+			// We could use a HashSet/BTreeSet for tracking the visited nodes but realistically,
```

Comment:
Why not just use a set right away?

Reply:
removed this comment. no longer needed because we allow cycles